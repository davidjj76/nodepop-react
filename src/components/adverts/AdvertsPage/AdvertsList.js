import React from 'react';
import T from 'prop-types';
import { Link } from 'react-router-dom';

import { advert } from '../propTypes';
import { capitalize } from '../../../utils/strings';

function Advert({ name, sale, price, tags }) {
  return (
    <div className="mx-auto cursor-pointer h-full hover:border-gray-400 transform transition-all duration-200 ease hover:-translate-y-1 shadow-sm w-72 max-w-full border border-gray-300 rounded-sm bg-white">
      <div className="w-full h-48 border-b border-gray-200">
        {sale ? 'Sell' : 'Buy'}
      </div>
      <div className="p-6">
        <div className="text-sm">
          <h3 className="font-bold text-base">{name}</h3>
          <div className="flex items-center text-green-400">{price} €</div>
          <p className="mt-1">{tags.map(capitalize).join(' · ')}</p>
        </div>
      </div>
    </div>
  );
}

Advert.propTypes = {
  ...advert,
};

function AdvertsList({ adverts }) {
  const renderAdvert = ({ id, ...advert }) => (
    <li key={id}>
      <Link to={`/adverts/${id}`}>
        <Advert {...advert} />
      </Link>
    </li>
  );

  return (
    <ul className="grid grid-cols-1 sm:grid-cols-3 xl:grid-cols-4 gap-6 mt-12">
      {adverts.map(renderAdvert)}
    </ul>
  );
}

AdvertsList.propTypes = {
  adverts: T.arrayOf(T.shape({ id: T.string.isRequired }).isRequired)
    .isRequired,
};

export default AdvertsList;
