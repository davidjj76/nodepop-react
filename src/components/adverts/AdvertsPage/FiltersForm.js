import T from 'prop-types';

import { Form, Input, FormConsumer } from '../../form';
import SelectTags from '../SelectTags';
import { RadioGroup, SelectRange } from '../../shared';
import { advert } from '../propTypes';
import { saleFilter } from './filters';

function FiltersForm({ initialFilters, defaultFilters, onFilter, prices }) {
  const handleResetClick = ({ setFormValue }) => () => {
    setFormValue(defaultFilters);
    onFilter(defaultFilters);
  };

  const min = Math.min(...prices);
  const max = Math.max(...prices);

  return (
    <Form initialValue={initialFilters} onSubmit={onFilter}>
      <p>Filters</p>
      <Input type="text" name="name" />
      <Input
        component={RadioGroup}
        name="sale"
        options={Object.values(saleFilter)}
      />
      <Input
        component={SelectRange}
        name="price"
        min={min}
        max={max}
        style={{ width: 400, margin: 24 }}
        marks={{ [min]: min, [max]: max }}
      />
      <Input component={SelectTags} name="tags" />
      <button type="submit">Filter</button>
      <FormConsumer>
        {form => <button onClick={handleResetClick(form)}>Reset</button>}
      </FormConsumer>
    </Form>
  );
}

const filtersProp = T.shape({
  ...advert,
  sale: T.oneOf(Object.keys(saleFilter)).isRequired,
  price: T.arrayOf(T.number.isRequired).isRequired,
});

FiltersForm.propTypes = {
  initialFilters: filtersProp.isRequired,
  defaultFilters: filtersProp.isRequired,
  onFilter: T.func.isRequired,
  prices: T.arrayOf(T.number.isRequired).isRequired,
};

export default FiltersForm;
