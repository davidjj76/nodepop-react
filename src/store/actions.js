import {
  getAdvertDetail,
  getAreAdvertsLoaded,
  getAreTagsLoaded,
} from './selectors';
import {
  AUTH_LOGIN_REQUEST,
  AUTH_LOGIN_SUCCESS,
  AUTH_LOGIN_FAILURE,
  AUTH_LOGOUT,
  ADVERTS_LOADED_REQUEST,
  ADVERTS_LOADED_SUCCESS,
  ADVERTS_LOADED_FAILURE,
  ADVERTS_CREATED_REQUEST,
  ADVERTS_CREATED_SUCCESS,
  ADVERTS_CREATED_FAILURE,
  ADVERTS_DETAIL_REQUEST,
  ADVERTS_DETAIL_SUCCESS,
  ADVERTS_DETAIL_FAILURE,
  ADVERTS_DELETED_REQUEST,
  ADVERTS_DELETED_SUCCESS,
  ADVERTS_DELETED_FAILURE,
  TAGS_LOADED,
  UI_RESET_ERROR,
} from './types';

const createRequestAction = type => () => ({
  type,
});

const createSuccessAction = type => payload => ({
  type,
  payload,
});

const createFailureAction = type => payload => ({
  type,
  payload,
  error: true,
});

export const authLoginRequest = createRequestAction(AUTH_LOGIN_REQUEST);
export const authLoginSuccess = createSuccessAction(AUTH_LOGIN_SUCCESS);
export const authLoginFailure = createFailureAction(AUTH_LOGIN_FAILURE);

export const loginAction = credentials => {
  return async function (dispatch, _getState, { api, history }) {
    dispatch(authLoginRequest());
    try {
      await api.auth.login(credentials);
      dispatch(authLoginSuccess());
      // Redirect
      const { from } = history.location.state || { from: { pathname: '/' } };
      history.replace(from);
    } catch (error) {
      dispatch(authLoginFailure(error));
    }
  };
};

export const authLogout = () => {
  return {
    type: AUTH_LOGOUT,
  };
};

export const logoutAction = () => {
  return async function (dispatch, _getState, { api }) {
    await api.auth.logout();
    dispatch(authLogout());
  };
};

export const advertsLoadedRequest = createRequestAction(ADVERTS_LOADED_REQUEST);
export const advertsLoadedSuccess = createSuccessAction(ADVERTS_LOADED_SUCCESS);
export const advertsLoadedFailure = createFailureAction(ADVERTS_LOADED_FAILURE);

export const advertsLoadAction = () => {
  return async function (dispatch, getState, { api }) {
    const advertsLoaded = getAreAdvertsLoaded(getState());
    if (advertsLoaded) {
      return;
    }

    dispatch(advertsLoadedRequest());
    try {
      const adverts = await api.adverts.getAdverts();
      dispatch(advertsLoadedSuccess(adverts));
      return adverts;
    } catch (error) {
      dispatch(advertsLoadedFailure(error));
    }
  };
};

export const advertsCreatedRequest = createRequestAction(
  ADVERTS_CREATED_REQUEST,
);
export const advertsCreatedSuccess = createSuccessAction(
  ADVERTS_CREATED_SUCCESS,
);
export const advertsCreatedFailure = createFailureAction(
  ADVERTS_CREATED_FAILURE,
);

export const advertsCreateAction = newAdvert => {
  return async function (dispatch, _getState, { api, history }) {
    dispatch(advertsCreatedRequest());
    try {
      const advert = await api.adverts.createAdvert(newAdvert);
      dispatch(advertsCreatedSuccess(advert));
      history.push(`/adverts/${advert.id}`);
      return advert;
    } catch (error) {
      dispatch(advertsCreatedFailure(error));
    }
  };
};

export const advertsDetailRequest = createRequestAction(ADVERTS_DETAIL_REQUEST);
export const advertsDetailSuccess = createSuccessAction(ADVERTS_DETAIL_SUCCESS);
export const advertsDetailFailure = createFailureAction(ADVERTS_DETAIL_FAILURE);

export const advertsDetailAction = advertId => {
  return async function (dispatch, getState, { api }) {
    const advertLoaded = getAdvertDetail(getState(), advertId);
    if (advertLoaded) {
      return;
    }

    dispatch(advertsDetailRequest());
    try {
      const advert = await api.adverts.getAdvert(advertId);
      dispatch(advertsDetailSuccess(advert));
      return advert;
    } catch (error) {
      dispatch(advertsDetailFailure(error));
    }
  };
};

export const advertsDeletedRequest = createRequestAction(
  ADVERTS_DELETED_REQUEST,
);
export const advertsDeletedSuccess = createSuccessAction(
  ADVERTS_DELETED_SUCCESS,
);
export const advertsDeletedFailure = createFailureAction(
  ADVERTS_DELETED_FAILURE,
);

export const advertsDeleteAction = advertId => {
  return async function (dispatch, _getState, { api, history }) {
    dispatch(advertsDeletedRequest());
    try {
      await api.adverts.deleteAdvert(advertId);
      dispatch(advertsDeletedSuccess(advertId));
      history.push('/');
    } catch (error) {
      dispatch(advertsDeletedFailure(error));
    }
  };
};

export const tagsLoaded = tags => {
  return {
    type: TAGS_LOADED,
    payload: tags,
  };
};

export const tagsLoadAction = () => {
  return async function (dispatch, getState, { api }) {
    if (getAreTagsLoaded(getState())) {
      return;
    }
    const tags = await api.adverts.getTags();
    dispatch(tagsLoaded(tags));
    return tags;
  };
};

export const resetError = () => {
  return {
    type: UI_RESET_ERROR,
  };
};
