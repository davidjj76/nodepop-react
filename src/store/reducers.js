export const initialState = {
  auth: false,
  adverts: {
    loaded: false,
    data: [],
  },
  ui: {
    loading: false,
    error: null,
  },
};

export const auth = (state = initialState.auth) => state;

export const adverts = (state = initialState.adverts) => state;

export const ui = (state = initialState.ui) => state;
