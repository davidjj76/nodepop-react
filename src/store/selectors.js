export const getIsLogged = state => state.auth;

export const getTags = state => state.tags;
export const getAreTagsLoaded = state => getTags(state).length > 0;

export const getUi = state => state.ui;
