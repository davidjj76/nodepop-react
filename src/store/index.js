import { applyMiddleware, combineReducers, createStore } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import thunk from 'redux-thunk';

import * as api from '../api';
import * as reducers from './reducers';

const configureStore = ({ preloadedState, history }) => {
  const middleware = [thunk.withExtraArgument({ api, history })];

  return createStore(
    combineReducers(reducers),
    preloadedState,
    composeWithDevTools(applyMiddleware(...middleware)),
  );
};

export default configureStore;
