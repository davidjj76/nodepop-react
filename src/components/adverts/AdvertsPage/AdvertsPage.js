import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Redirect } from 'react-router-dom';

import Layout from '../../layout';
import FiltersForm from './FiltersForm';
import AdvertsList from './AdvertsList';
import EmptyList from './EmptyList';
import storage from '../../../utils/storage';
import { defaultFilters, filterAdverts } from './filters';
import { getAdverts, getUi } from '../../../store/selectors';
import { advertsLoadAction } from '../../../store/actions';

const getFilters = () => storage.get('filters') || defaultFilters;
const saveFilters = filters => storage.set('filters', filters);

function AdvertsPage() {
  const { loading, error } = useSelector(getUi);
  const adverts = useSelector(getAdverts);
  const [filters, setFilters] = React.useState(getFilters);
  const dispatch = useDispatch();

  React.useEffect(() => {
    dispatch(advertsLoadAction());
  }, [dispatch]);

  React.useEffect(() => {
    saveFilters(filters);
  }, [filters]);

  if (error?.statusCode === 401) {
    return <Redirect to="/login" />;
  }

  const filteredAdverts = filterAdverts(adverts, filters);

  return (
    <Layout>
      {adverts.length > 0 && (
        <FiltersForm
          initialFilters={filters}
          defaultFilters={defaultFilters}
          prices={adverts.map(({ price }) => price)}
          onFilter={setFilters}
        />
      )}
      {filteredAdverts.length ? (
        <AdvertsList adverts={filteredAdverts} />
      ) : (
        <EmptyList advertsCount={adverts.length} />
      )}
    </Layout>
  );
}

export default AdvertsPage;
