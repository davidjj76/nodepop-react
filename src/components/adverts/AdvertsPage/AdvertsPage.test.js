import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';

import AdvertsPage from './AdvertsPage';
import { initialState } from '../../../store/reducers';
import { advertsLoadAction } from '../../../store/actions';

jest.mock('../../../store/actions');

describe('AdvertsPage', () => {
  let wrapper;
  const store = {
    getState: () => initialState,
    subscribe: () => {},
    dispatch: jest.fn(),
  };

  beforeEach(() => {
    wrapper = mount(
      <Provider store={store}>
        <Router>
          <AdvertsPage />
        </Router>
      </Provider>,
    );
  });

  test('should render snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should dispatch advertsLoadAction action', () => {
    expect(store.dispatch).toHaveBeenCalled();
    expect(advertsLoadAction).toHaveBeenCalled();
  });
});
