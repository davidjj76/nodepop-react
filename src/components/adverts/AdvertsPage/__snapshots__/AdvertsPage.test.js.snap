// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdvertsPage should render snapshot 1`] = `
<Provider
  store={
    Object {
      "dispatch": [MockFunction] {
        "calls": Array [
          Array [
            undefined,
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
      "getState": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <AdvertsPage>
        <Layout>
          <Header>
            <header>
              Nodepop React
              <nav>
                <ul>
                  <li>
                    <NavLink
                      activeClassName="selected"
                      isActive={[Function]}
                      to="/adverts"
                    >
                      <Link
                        aria-current={null}
                        to={
                          Object {
                            "hash": "",
                            "pathname": "/adverts",
                            "search": "",
                            "state": null,
                          }
                        }
                      >
                        <LinkAnchor
                          aria-current={null}
                          href="/adverts"
                          navigate={[Function]}
                        >
                          <a
                            aria-current={null}
                            href="/adverts"
                            onClick={[Function]}
                          >
                            Nodepop
                          </a>
                        </LinkAnchor>
                      </Link>
                    </NavLink>
                  </li>
                  <li>
                    <NavLink
                      activeClassName="selected"
                      isActive={[Function]}
                      to="/adverts/new"
                    >
                      <Link
                        aria-current={null}
                        to={
                          Object {
                            "hash": "",
                            "pathname": "/adverts/new",
                            "search": "",
                            "state": null,
                          }
                        }
                      >
                        <LinkAnchor
                          aria-current={null}
                          href="/adverts/new"
                          navigate={[Function]}
                        >
                          <a
                            aria-current={null}
                            href="/adverts/new"
                            onClick={[Function]}
                          >
                            New advert
                          </a>
                        </LinkAnchor>
                      </Link>
                    </NavLink>
                  </li>
                </ul>
              </nav>
              <Connect(AuthButton)>
                <AuthButton
                  isLogged={false}
                  onLogout={[Function]}
                >
                  <Link
                    to="/login"
                  >
                    <LinkAnchor
                      href="/login"
                      navigate={[Function]}
                    >
                      <a
                        href="/login"
                        onClick={[Function]}
                      >
                        Login
                      </a>
                    </LinkAnchor>
                  </Link>
                </AuthButton>
              </Connect(AuthButton)>
            </header>
          </Header>
          <main>
            <EmptyList
              advertsCount={0}
            >
              <div>
                <p>
                  No adverts here!
                </p>
                <Link
                  to="/adverts/new"
                >
                  <LinkAnchor
                    href="/adverts/new"
                    navigate={[Function]}
                  >
                    <a
                      href="/adverts/new"
                      onClick={[Function]}
                    >
                      Create the first advert
                    </a>
                  </LinkAnchor>
                </Link>
              </div>
            </EmptyList>
          </main>
          <Footer>
            <footer>
              @Keepcoding 2021
            </footer>
          </Footer>
        </Layout>
      </AdvertsPage>
    </Router>
  </BrowserRouter>
</Provider>
`;
